Introduce configuration "atLeastOne" for getting tags and flavors

Arnei opened the pull request:
My suggestion on how to fix #3530.

Most workflow operation handlers use `.getTagsAndFlavors` to parse tags and flavors (and if they don't they probably should). The `Configuration` arguments determine whether there should be zero, exactly one or an arbitrary amount of e.g. source flavors.

This commit introduces a new `Configuration` option called "atLeastOne", so that there for example must be at least one source flavor or more, but not zero.

This commit also introduces a new parameter list for `.getTagsAndFlavors` where the `Configuration` for both source tags and source flavors can be specified together. This means that the configuration will be applied to the sum of source tags and source flavors. I imagine this will mostly be used in combination with "atLeastOne", for WOHs that want at least one source flavor or tag specified, but don't particularly care which one.

### How to test this patch

Modify the segmentpreviews operation in the fast workflow with different combinations of source flavors and source tags. (Be sure to choose a presentation video long enough for segment-video to generate segments).

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

Arnei committed:
Introduce configuration "atLeastOne" for getting tags and flavors

My suggestion on how to fix #3530.

Most workflow operation handlers use `.getTagsAndFlavors` to parse tags and flavors
(and if they don't they probably should). The `Configuration` arguments determine
whether there should be zero, exactly one or an arbitrary amount of e.g. source flavors.

This commit introduces a new `Configuration` option called "atLeastOne", so that there
for example must be at least one source flavor or more, but not zero.

This commit also introduces a new parameter list for `.getTagsAndFlavors` where
the `Configuration` for both source tags and source flavors can be specified together.
I imagine this will mostly be used in combination with "atLeastOne", for WOHs that
want at least one source flavor or tag specified, but don't particularly care which one.

gregorydlogan commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/AbstractWorkflowOperationHandler.java:
```diff
@@ -498,21 +530,15 @@ protected ConfiguredTagsAndFlavors getTagsAndFlavors(WorkflowInstance workflow,
         break;
+      case many:
+        targetFlavorList = getFlavors(operation, TARGET_FLAVORS, TARGET_FLAVOR);
+        break;
```
```suggestion
        targetTagMap = parseTargetTagsByType(targetTagList);
        break;
```

gregorydlogan commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/AbstractWorkflowOperationHandler.java:
```diff
@@ -498,21 +530,15 @@ protected ConfiguredTagsAndFlavors getTagsAndFlavors(WorkflowInstance workflow,
+        targetFlavorList = getFlavors(operation, TARGET_FLAVORS, TARGET_FLAVOR);
+        if (targetTagList.isEmpty()) {
+          throw new WorkflowOperationException("Configuration key '" + TARGET_FLAVORS + "' or '" + TARGET_FLAVOR + "' must be set");
         }
```
```suggestion
        }
        targetTagMap = parseTargetTagsByType(targetTagList);
```

Arnei replied:
Could you elaborate?

gregorydlogan reviewed (changes requested):
This fixes the ~compile~test issues.  Still working on testing, but it looks like it should work.

jduehring commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SegmentPreviewsWorkflowOperationHandler.java:
```diff
@@ -192,7 +192,7 @@ private WorkflowOperationResult createPreviews(final MediaPackage mediaPackage,
     ConfiguredTagsAndFlavors tagsAndFlavors = getTagsAndFlavors(wi,
-        Configuration.many, Configuration.one, Configuration.many, Configuration.one);
+        Configuration.atLeastOne, Configuration.many, Configuration.one);
     MediaPackageElementFlavor sourceVideoFlavor = tagsAndFlavors.getSingleSrcFlavor();
```
Line 196 still leads to an error if no `source-flavor` is specified, even though one `source-tag` should be sufficient.

I suggest to handle it like the tags and get a list of all flavors (including empty list, if no flavors exist) and adding these to the TrackSelector.

Arnei replied:
Thanks for the review. I applied your suggestions.

jduehring commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SegmentPreviewsWorkflowOperationHandler.java:
```diff
@@ -192,7 +192,7 @@ private WorkflowOperationResult createPreviews(final MediaPackage mediaPackage,
     ConfiguredTagsAndFlavors tagsAndFlavors = getTagsAndFlavors(wi,
-        Configuration.many, Configuration.one, Configuration.many, Configuration.one);
+        Configuration.atLeastOne, Configuration.many, Configuration.one);
     MediaPackageElementFlavor sourceVideoFlavor = tagsAndFlavors.getSingleSrcFlavor();
```
```suggestion
    List<MediaPackageElementFlavor> sourceVideoFlavors = tagsAndFlavors.getSrcFlavors();
```

jduehring commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SegmentPreviewsWorkflowOperationHandler.java:
```suggestion
    for (MediaPackageElementFlavor flavor : sourceVideoFlavors) {
      trackSelector.addFlavor(flavor);
    }
```

jduehring reviewed (commented):
I tried to test different combinations of `source-flavors` and `source-tags` as you suggested. Seemed to mostly work fine.
But only specifiying one `source-tag` still leads to an error, even though this should work if I understood the initial issue correctly. 

We should also probably check the docs and describe in the WOHs how many tags/flavors are expected/required in each operation.

Arnei committed:
Fix exception if no source flavor speciied for SegmentPreviesWorkflow

It should only if fail if there are no source flavors AND no source tags, but it also failed when there were source tags but no source flavors. This should fix that.

Co-authored-by: Jonas Dühring <32971521+jduehring@users.noreply.github.com>

Arnei committed:
Add note about source flavors/tags to docs for segment previews WOH

This changed to require only at least one source flavor or source tag.

Arnei committed:
Fix misnamed variable
